### ISSUE TITLE
Placeholder issue for GitHub PR scenario

### ISSUE THREAD

**Issue by arvindsv** [gocd/gocd#2470]
This needs me to put in more details and results from further investigation. Scenario goes something like this:

1. Configure the PR plugin
2. Open a PR for a commit.

3. 

  a. Open another PR for the same commit (or maybe merge it into another branch and push that branch and submit a PR for that).

- OR -

  a. Submit a PR for a commit
  b. Let build finish
  c. Submit a commit on the same PR
  d. Let build finish
  e. `git reset HEAD^1` to go back to previous commit, force push
  f. warning in gocd.

The plugin responds with the old commit and GoCD server 16.6.0 and beyond has some code to reject duplicate commits. The PR build does not trigger.

**Comment by arvindsv** [comment]
Don't know yet if this a PR plugin issue or something in core. That's why this issue is here.

**Comment by stania1** [comment]
More detailed steps to help replicate the issue:

1. Configure a pipeline with the PR plugin (Github material)
2. Create a branch, push a commit, open a PR for that branch (PR 1). You may have to close this PR right away.
3. Wait for a while. Let Go pick up that PR and commit. Only proceed when the material in the DB already has PR 1 and that commit.
4. Open another PR (PR 2). Make sure that the PR ID is _actually_ 1 above the previous one (e.g. if the old one is 2006, this one should be 2007). I think this number is essential, see **(*)** below.
5. Let Go + the Github PR pick up this commit. Actually, it won't and will pretty much refuse to pick up subsequent commits. (I wrote "pretty much" because in the early hours of noticing this problem I saw some, but very few, commits still getting picked up, before it stopped completely. Not sure why it did that yet.) What I saw happening in the UI:
```
Modification check failed for material: [url=https://github.com/bla/repo.git] [Jul-20 12:16:10]
Cannot save modification: All modifications already exist in db: [commit-SHA]
```

In `go-server.log`:

```
2016-07-19 13:10:43,795  INFO [94@MessageListener for MaterialUpdateListener] ScmMaterialUpdater:63 - [Material Update] Found '1' modifications for material ''PluggableSCMMaterial{[url=https://github.com/bla/repo.git]}'' with flyweight '...' using working directory '/home/go-server/pipelines/flyweight/...'
2016-07-19 13:10:43,796 ERROR [94@MessageListener for MaterialUpdateListener] MaterialRepository:818 - Cannot save modification:
java.lang.RuntimeException: All modifications already exist in db: [commit-SHA]
        at com.thoughtworks.go.server.persistence.MaterialRepository.checkAndRemoveDuplicates(MaterialRepository.java:852)
        at com.thoughtworks.go.server.persistence.MaterialRepository.saveModifications(MaterialRepository.java:814)
        at com.thoughtworks.go.server.materials.ScmMaterialUpdater.insertLatestOrNewModifications(ScmMaterialUpdater.java:66)
        at com.thoughtworks.go.server.materials.PluggableSCMMaterialUpdater.insertLatestOrNewModifications(PluggableSCMMaterialUpdater.java:61)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.insertLatestOrNewModifications(MaterialDatabaseUpdater.java:155)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterialWithNewRevisions(MaterialDatabaseUpdater.java:147)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater$2.doInTransaction(MaterialDatabaseUpdater.java:110)
        at com.thoughtworks.go.server.transaction.TransactionCallback.doWithExceptionHandling(TransactionCallback.java:24)
        at com.thoughtworks.go.server.transaction.TransactionTemplate$3.doInTransaction(TransactionTemplate.java:53)
        at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:130)
        at com.thoughtworks.go.server.transaction.TransactionTemplate.executeWithExceptionHandling(TransactionTemplate.java:49)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterial(MaterialDatabaseUpdater.java:107)
        at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:50)
        at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:30)
        at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.runImpl(JMSMessageListenerAdapter.java:65)
        at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.run(JMSMessageListenerAdapter.java:50)
        at java.lang.Thread.run(Thread.java:745)
2016-07-19 13:10:43,797  WARN [94@MessageListener for MaterialUpdateListener] MaterialDatabaseUpdater:121 - [Material Update] Modification check failed for material: [url=https://github.com/bla/repo.git]
java.lang.RuntimeException: Cannot save modification: All modifications already exist in db: [commit-SHA]
        at com.thoughtworks.go.server.persistence.MaterialRepository.saveModifications(MaterialRepository.java:819)
        at com.thoughtworks.go.server.materials.ScmMaterialUpdater.insertLatestOrNewModifications(ScmMaterialUpdater.java:66)
        at com.thoughtworks.go.server.materials.PluggableSCMMaterialUpdater.insertLatestOrNewModifications(PluggableSCMMaterialUpdater.java:61)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.insertLatestOrNewModifications(MaterialDatabaseUpdater.java:155)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterialWithNewRevisions(MaterialDatabaseUpdater.java:147)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater$2.doInTransaction(MaterialDatabaseUpdater.java:110)
        at com.thoughtworks.go.server.transaction.TransactionCallback.doWithExceptionHandling(TransactionCallback.java:24)
        at com.thoughtworks.go.server.transaction.TransactionTemplate$3.doInTransaction(TransactionTemplate.java:53)
        at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:130)
        at com.thoughtworks.go.server.transaction.TransactionTemplate.executeWithExceptionHandling(TransactionTemplate.java:49)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterial(MaterialDatabaseUpdater.java:107)
        at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:50)
        at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:30)
        at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.runImpl(JMSMessageListenerAdapter.java:65)
        at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.run(JMSMessageListenerAdapter.java:50)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: All modifications already exist in db: [2cf2643807a9bafd7adaa5696dcb1dd229a6d0ea]
        at com.thoughtworks.go.server.persistence.MaterialRepository.checkAndRemoveDuplicates(MaterialRepository.java:852)
        at com.thoughtworks.go.server.persistence.MaterialRepository.saveModifications(MaterialRepository.java:814)
        ... 15 more
```

What seems to have happened:
Even though Github PR plugin sees a lot of changes, it picks just [one](https://github.com/ashwanthkumar/gocd-build-github-pull-requests/blob/master/src/main/java/in/ashwanthkumar/gocd/github/GitHubPRBuildPlugin.java#L212) (and it does so based on whatever ordering `git show-ref` gives, which is why I think the PR numbers need to be sequential **(*)**) to tell the GoCD server about. This one commit that got picked up is already in the database as it was also part of a previous PR. GoCD refuses to save the modification as there was only 1 modification and it was a duplicate.

Workaround ideas:
1. If PR 2 has NOT been merged, push another commit to its underlying branch and it should be ok.
2. If PR 2 HAS been merged, the only workaround I could think of is to hack the PR code a little to skip PR 2 when returning the list of new modifications to GoCD. Pushing a commit will not do anything as Github ignores changes to that branch after the PR has been merged.

**Comment by stania1** [comment]
This just happened a second time for us. Closing a PR and opening another one which contains the same commit isn't that uncommon apparently.

**Comment by tomzo** [comment]
I have also seen this error `Cannot save modification: All modifications already exist in db:` and spent few hours trying to reproduce this on development server after importing db from production server. 
However that **Go server has never used Github PR plugin, this happened on standard git poller.**
No luck reproducing it though. In production error happened only on one repository. It was a configuration repo and a material in a pipeline. It seems to have been caused by having first a config on `ci` branch
```xml
    <config-repo plugin="json.config.plugin">
      <git url="git@example.com" branch="ci" />
    </config-repo>
```
Then branch was changed to `master`, either by specifying `branch="master"` or skipping `branch` tag at all. I don't know which was it. 
Most likely both ci and master had same commits at that time.
Really weird part is that this had been happening on every server restart and only in that one repository, which is not special in any way compared to dozens of others. Quick workaround was to rewrite last commit, which solved it until next restart.
I decided that something in db got screwed up and ended up renaming the URL of material, so Go sees it as a new material, which has "solved it".

I don't know is this is any helpful, but maybe the problem is somewhere in MDU, I doubt that is github PR poller related. I cannot reproduce this, this happened on server which was running on so many unofficial versions that it is possible that it was some temporary issue.

@arvindsv please elaborate what you meant by this:

> The plugin responds with the old commit and GoCD server 16.6.0 and beyond has some code to reject duplicate commits.

So there is a feature to reject duplicate commits? If same commit exists in 2 branches then is it a *duplicate commit*?

**Comment by stania1** [comment]
Yeah, I need to clarify this a little. I have seen this happen without the Github PR plugin too, but haven't investigated or looked into reproducing it. However, when it happens to a pipeline that uses the standard git poller, it usually resolves itself - the next commit that comes (or a rewrite of the last commit) will unblock the pipeline.

When this happens on a pipeline with the Github PR plugin, it's a little trickier to resolve because of the plugin's behavior as described on my previous post.

**Comment by zabil** [comment]
@stania1 what version of go.cd are you using? [16.5.0+](https://www.go.cd/releases/index.html?#16.5.0) fixed issues with git rebase

@tomzo @arvindsv is talking about this
https://github.com/gocd/gocd/blob/2584deab37edc05ffd02532370dc635bd1dc5a05/server/src/com/thoughtworks/go/server/persistence/MaterialRepository.java#L814

Depending on how you rebase git gocd's the git poller can return modifications/commits that's already inserted into the db (as a part of earlier polls). So a check was added to not insert duplicate modifications.

**Comment by stania1** [comment]
@zabil I'm using 16.6.0.

**Comment by rajiesh** [comment]
Verified on GoCD version 16.11.0 (4185-e7acc53aac8bb5ad2a7f306ca8f3e752855d844b) working as expected

**Comment by arvindsv** [comment]
For the record, the PR that should have fixed/hidden this is https://github.com/gocd/gocd/pull/2803.
